kubernetes/e2e_test/test_client.py

```diff
@@ -55,9 +55,12 @@ class TestClient(unittest.TestCase):
     def setUpClass(cls):
         cls.config = base.get_e2e_configuration()
 
+    def setUp(self):
+        self.client = api_client.ApiClient(configuration=self.config,
+                                           prefered_content_type="application/strategic-merge-patch+json")
```

Comment:
We can use it as a default to make it backward compatible.

Reply:
Assume you meant to have it as default in the ApiClient?

Reply:
IIUC, defaulting to one content type doesn't make it backward compatible, since the old behavior was picking one of the two content types based on the `body` type. We have to keep the old behavior if we want to avoid breaking the old clients

Reply:
IMO backward compatibility is not possible, because current behavior doesn't work for custom resources. If we assume that strategic-merg-patch is more common we can set is as a default. Other clients will have to change the preferred content type only...

Reply:
> IMO backward compatibility is not possible, because current behavior doesn't work for custom resources

the current behavior allows JSON patch for custom resources, right? Can we keep the following behavior if `prefered_content_type` is None? 
1. for native types, a client can send `json-patch+json` if the body is a list, or `strategic-merge-patch+json` if the body is a map
2. for custom resources, a client can send `json-patch+json` (a list) only (apiserver doesn't support strategic merge patch for custom resources)

In future, we need to allow a client to send `merge-patch+json` and `apply-patch+yaml` for both native types and custom resources as well

Reply:
Sure, I prepare changes for such approach.

I see one drawbacks - we'll keep the "magic" logic in rest.py and it won't be such generic to add it to the open-generator. But let's discuss it later with the code.

Thanks.

Reply:
if the `select_header_content_type` takes the body and method as well then it can be intelligent about it's choice. We should be prepared for that custom resources will support strategic merge in the future when such strategy is defined in the CRD. So it is still a good idea to introduce the option to set the preferred type. 

Note that the previous supported type for custom resources was `application/merge-patch+json` and not `json-patch+json` which was newly added. So if you should have a default for custom objects you should select `application/merge-patch+json` if you want backward compatibility.

Reply:
> Note that the previous supported type for custom resources was application/merge-patch+json and not json-patch+json which was newly added. So if you should have a default for custom objects you should select application/merge-patch+json if you want backward compatibility

Ah yes, you're right. Thanks!

Reply:
Now I’m leaning towards to expose content_type to each request (for patch mainly) to bypass the select_content_type function. This function now accepts "method" and "body" to try to select the better header. K8s introduces strategic-merge-patch which is k8s specific and this part has to be patched - instead of patching rest.py. Let me know what you think. Thanks.